composer/tools/copy_environment.py

```diff
@@ -0,0 +1,761 @@
+                sql_client, prev_sql_project, prev_sql_instance
+            ),
+        )
+        print("Exporting database from old Environment.")
+        export_data(
+            sql_client,
+            prev_sql_project,
+            prev_sql_instance,
+            gcs_db_dump_bucket.name,
+            "db_dump.sql",
+        )
+        print("Obtaining fernet keys for Composer Environments.")
+        old_fernet_key = get_fernet_key(existing_env)
+        new_fernet_key = get_fernet_key(new_env)
+        print("Preparing database import to new Environment.")
+        import_data(
+            sql_client,
+            gke_service_account_key,
+            new_sql_project,
+            new_sql_instance,
+            gcs_db_dump_bucket.name,
+            "db_dump.sql",
+            old_fernet_key,
+            new_fernet_key,
+        )
+    finally:
+        if gke_service_account_key:
+            print("Deleting temporary service account key.")
+            delete_service_account_key(
+                iam_client,
+                project,
+                gke_service_account_name,
+                gke_service_account_key,
+            )
+        if gcs_db_dump_bucket:
+            print("Deleting temporary Cloud Storage bucket.")
+            delete_bucket(gcs_db_dump_bucket)
+
+
+def copy_gcs_bucket(existing_env, new_env):
+    print("Starting to transfer Cloud Storage artifacts.")
+    existing_bucket = existing_env["config"]["dagGcsPrefix"][:-4]
+    new_bucket = new_env["config"]["dagGcsPrefix"][:-4]
+    for subdir in ["dags", "plugins", "data", "logs"]:
+        subprocess.call(
+            [
+                "gsutil",
+                "-m",
+                "cp",
+                "-r",
+                existing_bucket + subdir + "/*",
+                new_bucket + subdir,
+            ]
+        )
+
+
+def clone_environment(
+    project,
+    location,
+    existing_env_name,
+    new_env_name,
+    running_as_service_account,
+    overrides,
+):
+    default_credentials, _ = google.auth.default(scopes=DEFAULT_SCOPES)
+
+    composer_api_version = "v1"
+
+    if overrides["betaApi"]:
+        composer_api_version = "v1beta1"
+
+    composer_client = discovery.build(
+        "composer", composer_api_version, credentials=default_credentials
+    )
+
+    existing_env = get_composer_env(
+        composer_client, project, location, existing_env_name
+    )
+    create_composer_env_if_not_exist(
+        composer_client,
+        existing_env,
+        project,
+        location,
+        new_env_name,
+        overrides,
+    )
+    new_env = get_composer_env(
+        composer_client, project, location, new_env_name
+    )
+    update_pypi_packages(composer_client, existing_env, new_env)
+    new_env = get_composer_env(
+        composer_client, project, location, new_env_name
+    )
+    copy_database(project, existing_env, new_env, running_as_service_account)
+    copy_gcs_bucket(existing_env, new_env)
+    print(
```

Comment:
Prefer string format over implicit concatentation:

```
print('{}{}{}'.format(
    'Composer Environment copy completed. Please check the new ',
    'environment is correct and delete the old one to avoid ',
    'incurring additional costs.'))
```